extensions/lifecycle/lifecycleProducer/LifecycleProducer.js

```diff
@@ -0,0 +1,386 @@
+'use strict'; // eslint-disable-line
+
+const fs = require('fs');
+const async = require('async');
+const AWS = require('aws-sdk');
+const http = require('http');
+const { Logger } = require('werelogs');
+const { errors } = require('arsenal');
+
+const BackbeatProducer = require('../../../lib/BackbeatProducer');
+const BackbeatConsumer = require('../../../lib/BackbeatConsumer');
+const VaultClientCache = require('../../../lib/clients/VaultClientCache');
+const safeJsonParse = require('../util/safeJsonParse');
+
+const CONSUMER_FETCH_MAX_BYTES = 5000020;
+
+/**
+ * @class LifecycleProducer
+ *
+ * @classdesc Handles consuming entries from the bucket topic and sending those
+ * entries to be processed by the lifecycle task. Once processed, the entry is
+ * sent to the object topic for another consumer to perform the lifecycle
+ * operation (e.g., expiration) on the object. If another listing is required on
+ * the bucket (i.e., the listing is truncated), an entry is pushed to the bucket
+ * topic for the next round of processing.
+ */
+class LifecycleProducer {
+    /**
+     * Constructor of LifecycleProducer
+     * @param {Object} zkConfig - zookeeper config
+     * @param {Object} lcConfig - lifecycle config
+     * @param {Object} s3Config - s3 config
+     * @param {String} s3Config.host - host ip
+     * @param {String} s3Config.port - port
+     * @param {Object} authConfig - auth config
+     * @param {String} [authConfig.account] - account name
+     * @param {Object} [authConfig.vault] - vault details
+     * @param {String} authConfig.vault.host - vault host ip
+     * @param {number} authConfig.vault.port - vault port
+     * @param {number} authConfig.vault.adminPort - vault admin port
+     * @param {String} transport - http or https
+     */
+    constructor(zkConfig, lcConfig, s3Config, authConfig, transport) {
+        this._log = new Logger('Backbeat:LifecycleProducer');
+        this._zkConfig = zkConfig;
+        this._lcConfig = lcConfig;
+        this._authConfig = authConfig;
+        this._s3Endpoint = `${transport}://${s3Config.host}:${s3Config.port}`;
+        this._transport = transport;
+        this._bucketProducer = null;
+        this._objectProducer = null;
+        this.accountCredsCache = {};
+
+        // The task scheduler for processing lifecycle tasks concurrently.
+        this._internalTaskScheduler = async.queue((ctx, cb) => {
+            const { task, rules, value } = ctx;
+            return task.processBucketEntry(rules, value, cb);
+        }, this._lcConfig.producer.concurrency);
+
+        // Listen for errors from any task being processed.
+        this._internalTaskScheduler.drain(err => {
+            if (err) {
+                this._log.error('error occured when during task processing', {
```

Comment:
trivial, but typo including `when during`. should use either one or the other

Reply:
Updated.